File: promotion-jvm/src/main/java/com/github/hippoom/xpro/domain/Order.java
package com.github.hippoom.xpro.domain;

import java.util.List;

public class Order {
	private int discount;
	private List<Item> items;

	public Order(List<Item> items) {
		this.items = items;
	}

	public int cost() {
		return origin() - discount;
	}

	public int origin() {
		int cost = 0;
		for (Item item : items) {
			cost += item.subtotal();
		}
		return cost;
	}

	public void apply(Coupon coupon) {
		coupon.applyTo(this);
	}

	public void apply(int applied) {
		this.discount = applied;
	}

	public List<Item> items() {
		return items;
	}
}

File: promotion-ruby/src/main/promotion/Order.rb
class Order
  def initialize items
    @items = items
  end

  def cost
    return origin - (@applied || 0)
  end

  def origin
    sum = 0
    @items.each {|item|
      sum += item.subtotal
    }
    return sum
  end

  def apply coupon
    @applied = coupon.apply_to self unless coupon.nil?
  end
end